### ISSUE TITLE
Check that FA2 speeds up training with SmolLM2-135M YAML

### ISSUE THREAD

**Issue by vishalbakshi** [LocalResearchGroup/llm-foundry#25]
@galopyz and I have both received the following warning:

```
Flash Attention 2.0 only supports torch.float16 and torch.bfloat16 dtypes, but the current dype in LlamaForCausalLM is torch.float32. You should run training or inference using Automatic Mixed-Precision via the `with torch.autocast(device_type='torch_device'):` decorator, or load the model with the `torch_dtype` argument. Example: `model = AutoModel.from_pretrained("openai/whisper-tiny", attn_implementation="flash_attention_2", torch_dtype=torch.float16)`
```
Even though our training YAML contains:

```
# System
seed: ${variables.global_seed}
device_eval_batch_size: 16
device_train_microbatch_size: 16
precision: amp_bf16
```

I have ran [a very small/rough experiment](https://github.com/LocalResearchGroup/llm-foundry/issues/14) to see if using FA2 speeds up training. We should run a longer experiment to ensure it does.

**Comment by warner-benjamin** [comment]
Training in mixed precision (amp_fp16 or amp_bf16) should correctly convert the FA inputs to fp16/bf16 ([except maybe on the latest version of transformers](https://github.com/huggingface/transformers/issues/35945), which is why we are not using it. If we are using an Ampere card, we probably want to be training in amp_bf16 by default. (Or maybe "quantize" the base weights to bf16 and use amp for the lora adapters)

For our own implementation of Llama/Qwen, we'll want to follow the [ModernBERT Flash Attention](https://github.com/huggingface/transformers/blob/6b550462139655d488d4c663086a63e98713c6b9/src/transformers/models/modernbert/modular_modernbert.py#L550) code to handle non-fp16/bf16 inputs.

**Comment by galopyz** [comment]
I ran some experiments with FA2 vs. No-FA2 on smolm2-135m using modal with l4 following @vishalbakshi 's #14 . I only paid attention to the speed of the training. I did not check the loss. 

With FA2, I could use batch size of 4, but No-FA2 could only use batch size of 1. 


The following is the runs on l4 GPU on modal on 100 batch and different batch sizes (bs):
| options | duration (min:sec) |
| --- | --- |
| FA2, 4 bs | 8:24 |
| FA2, 1 bs | 8:35 |
| No-FA2, 1 bs | 32:49 |

Here is one with 500 batch:
| options | duration (min:sec) |
| --- | --- |
| FA2, 4 bs | 12:18 |
| FA2, 1 bs | 13:18 |
| No-FA2, 1 bs | 37:45 |

It's a bit strange that durations for FA2, 1 bs are very similar between 100 batch and 500 batch. I also found out that inference took a significant more time without FA2. 
These runs are done using amp_bf16.

I can do more testing with 1000 batches or do more runs with same options. Or we can just use FA2 and move on.

**Comment by vishalbakshi** [comment]
@galopyz thanks for sharing this I think this is strong evidence that FA2 is being used. I agree that it is curious how bs=1 takes the same time whether it's 100ba or 500ba. Can you share the code you are using for both scenarios and I can see if I can recreate the result? Also, for duration are you looking at AIM metrics or measuring it some other way?

**Comment by galopyz** [comment]
@vishalbakshi sure. This is from AIM metrics from Mat's AIM server. 

Basically, from the modal training, I changed `train_cmd`, `BATCH_SIZE`, and commented out couple of the last lines:
Here is train_cmd. I changed experiment name, `max_duration`, `eval_interval`, and `model.use_flash_attention_2` attributes:
```py
 train_cmd = [
        "composer",
        "train/train.py",
        "train/yamls/pretrain/smollm2-135m.yaml",  # Updated YAML path
        "loggers.aim.experiment_name=galopy_smollm2_135m_modal_FA2_100ba_1bs",  # experiment name
        f"variables.data_local={DATASETS_VOLUME_MOUNT_PATH}/c4_small",
        "train_loader.dataset.split=train_small",
        "eval_loader.dataset.split=val_small",
        "max_duration=100ba",
        "eval_interval=100ba",
        f"save_folder={save_folder}",  # Updated model name
        f"device_eval_batch_size={BATCH_SIZE}",  # Added batch size settings # 20 for h100 4 for l4
        f"device_train_microbatch_size={BATCH_SIZE}",
        f"global_train_batch_size={BATCH_SIZE}",
        # "model.use_flash_attention_2=false" # No FA2
    ]
```
I commented out these lines from the training to only focus on training and eval.
```py
    # view_model_checkpoints.remote()
    # time.sleep(1)
    # convert_model_to_hf.remote(model_path)
    # time.sleep(1)
    # evaluate_model.remote(model_path+"/hf")
    # time.sleep(1)
    # generate_responses.remote(model_path+"/hf")
```

Here is the whole thing:
```py
# INSTRUCTIONS
# Create volumes (one time via cli) - run: `modal volume create lrg-datasets` and `modal volume create lrg-model-checkpoints`
#TODO:
# - Probably want to put local aim repo in the folder with the model checkpoints
# - Fix duplicate container issue


import modal
from modal import Image, App, Secret, Volume

import pathlib, datetime

PYTHON_PATH = "/opt/conda/envs/llm-foundry/bin/python"
TRAINING_GPU = "l4" # "a10g" "h100" # "l4"
BATCH_SIZE = 1 # 20 for h100 4 for l4
DATASET_BASE_PATH = "/datasets"
DATASETS_VOLUME = Volume.from_name("lrg-datasets")
DATASETS_VOLUME_MOUNT_PATH = pathlib.Path("/datasets")
MODEL_CHECKPOINT_VOLUME = Volume.from_name("lrg-model-checkpoints")
MODEL_CHECKPOINT_VOLUME_MOUNT_PATH = pathlib.Path("/model-checkpoints")

app = App("quick-start")

# Build image from local Dockerfile
image = Image.from_dockerfile("Dockerfile")

@app.function(gpu=TRAINING_GPU, image=image, timeout=3600, secrets=[Secret.from_name("LRG")],
             concurrency_limit=1)
def get_stats():
    import subprocess
    
    # Use the correct Python interpreter for imports
    import_check = subprocess.run(
        [PYTHON_PATH, "-c", "import flash_attn; print(flash_attn.__version__)"],
        capture_output=True,
        text=True
    )
    print(f"Flash Attention version: {import_check.stdout}")

    # Run nvidia-smi to check GPU status
    nvidia_smi = subprocess.run(['nvidia-smi'], capture_output=True, text=True)
    print("NVIDIA-SMI Output:")
    print(nvidia_smi.stdout)
    if nvidia_smi.stderr:
        print("NVIDIA-SMI Errors:", nvidia_smi.stderr)


@app.function(gpu=TRAINING_GPU, image=image, timeout=3600, secrets=[Secret.from_name("LRG")], 
              volumes={DATASETS_VOLUME_MOUNT_PATH: DATASETS_VOLUME},
              concurrency_limit=1)
def convert_c4_small_dataset():
    import subprocess
    import os
    
    # Change to llm-foundry/scripts directory at the start
    os.chdir("/llm-foundry/scripts")
    print(f"Working directory: {os.getcwd()}")
    
    # Step 1: Convert C4 dataset
    print("Converting C4 dataset...")
    data_prep_cmd = [
        PYTHON_PATH,  # Use the correct Python interpreter
        "data_prep/convert_dataset_hf.py",
        "--dataset", "allenai/c4",
        "--data_subset", "en",
        "--out_root", f"{DATASET_BASE_PATH}/c4_small",
        "--splits", "train_small", "val_small",
        "--concat_tokens", "2048",
        "--tokenizer", "HuggingFaceTB/SmolLM2-135M",
        "--eos_text", "<|endoftext|>"
    ]
    result = subprocess.run(data_prep_cmd, capture_output=True, text=True)
    print(result.stdout)
    if result.stderr:
        print("Data prep errors:", result.stderr)
    
    DATASETS_VOLUME.commit()

@app.function(gpu=TRAINING_GPU, image=image, timeout=3600, secrets=[Secret.from_name("LRG")], 
              volumes={MODEL_CHECKPOINT_VOLUME_MOUNT_PATH: MODEL_CHECKPOINT_VOLUME},
              concurrency_limit=1)
def view_model_checkpoints(save_folder: str=None):
    import os
    print("\nModel checkpoint files and sizes:")
    for folder in ([save_folder] if save_folder else os.listdir(MODEL_CHECKPOINT_VOLUME_MOUNT_PATH)):
        folder = os.path.join(MODEL_CHECKPOINT_VOLUME_MOUNT_PATH, folder)
        if os.path.isdir(folder):
            for filename in os.listdir(folder):
                filepath = os.path.join(folder, filename)
                if os.path.isfile(filepath):
                    size_mb = os.path.getsize(filepath) / (1024 * 1024)
                    print(f"{os.path.join(folder, filename)}: {size_mb:.2f} MB")


# @app.function(gpu=TRAINING_GPU, image=image, timeout=6*3600, secrets=[Secret.from_name("LRG")], 
#               volumes={MODEL_CHECKPOINT_VOLUME_MOUNT_PATH: MODEL_CHECKPOINT_VOLUME,
#                        DATASETS_VOLUME_MOUNT_PATH: DATASETS_VOLUME})
def train_model(run_ts: str):
    import subprocess
    import os
    
    # Change to llm-foundry/scripts directory at the start
    os.chdir("/llm-foundry/scripts")
    print(f"Working directory: {os.getcwd()}")
    
    # Step 2: Train the model
    print("\nTraining model...")
    save_folder = f"{MODEL_CHECKPOINT_VOLUME_MOUNT_PATH}/smollm2-135m-{run_ts}"
    
    train_cmd = [
        "composer",
        "train/train.py",
        "train/yamls/pretrain/smollm2-135m.yaml",  # Updated YAML path
        "loggers.aim.experiment_name=galopy_smollm2_135m_modal_FA2_100ba_1bs",
        f"variables.data_local={DATASETS_VOLUME_MOUNT_PATH}/c4_small",
        "train_loader.dataset.split=train_small",
        "eval_loader.dataset.split=val_small",
        "max_duration=100ba",
        "eval_interval=100ba",
        f"save_folder={save_folder}",  # Updated model name
        f"device_eval_batch_size={BATCH_SIZE}",  # Added batch size settings # 20 for h100 4 for l4
        f"device_train_microbatch_size={BATCH_SIZE}",
        f"global_train_batch_size={BATCH_SIZE}",
        # "model.use_flash_attention_2=false" # No FA2
    ]
    result = subprocess.run(train_cmd, capture_output=True, text=True)
    print(result.stdout)
    print(f'Training complete for {run_ts}')
    print(f'Model checkpoints saved to {save_folder}')

    MODEL_CHECKPOINT_VOLUME.commit()

    # Print checkpoint file sizes
    view_model_checkpoints.remote(save_folder)
    
    if result.stderr:
        print("Training errors:", result.stderr)
    if result.returncode != 0:
        raise Exception(f"Training failed with exit code {result.returncode}\nStderr: {result.stderr}")
    return str(save_folder)

def run_aim_server():
    import subprocess
    import os
    
    os.chdir("/llm-foundry/scripts")
    print("Initializing Aim...")
    subprocess.run(["aim", "init"], check=True)
    
    process = subprocess.Popen(
        ["aim", "up", "--host", "0.0.0.0", "--port", "43800"],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE
    )
    
    return process


@app.function(gpu=TRAINING_GPU, image=image, timeout=12*3600, secrets=[Secret.from_name("LRG")],
              volumes={MODEL_CHECKPOINT_VOLUME_MOUNT_PATH: MODEL_CHECKPOINT_VOLUME,
                      DATASETS_VOLUME_MOUNT_PATH: DATASETS_VOLUME},
              concurrency_limit=1)
def train_with_aim(run_ts: str):
    import subprocess, time

    with modal.forward(43800) as tunnel:
        print(f"\nAim server available at: {tunnel.url}")
        model_path = None
        aim_task = run_aim_server()
        time.sleep(5)
    
        try:
            model_path = train_model(run_ts)

        finally:
            aim_task.terminate()
            try:
                aim_task.wait(timeout=5)
            except subprocess.TimeoutExpired:
                aim_task.kill()
    
    return model_path

@app.function(gpu=TRAINING_GPU, image=image, timeout=3600, secrets=[Secret.from_name("LRG")],
              volumes={MODEL_CHECKPOINT_VOLUME_MOUNT_PATH: MODEL_CHECKPOINT_VOLUME},
              concurrency_limit=1)
def convert_model_to_hf(checkpoint_path: str):
    """
    Convert a model checkpoint to a HuggingFace format.
    """
    import subprocess, os
    from pathlib import Path

    os.chdir("/llm-foundry/scripts")
    print(f"Working directory: {os.getcwd()}")

    composer_checkpoint_path = Path(MODEL_CHECKPOINT_VOLUME_MOUNT_PATH)/checkpoint_path
    if composer_checkpoint_path.is_dir():
        composer_checkpoint_path = composer_checkpoint_path / "latest-rank0.pt"
    hf_output_path = composer_checkpoint_path.parent / "hf"

    print("\nConverting model to HuggingFace format...")
    convert_cmd = [
        PYTHON_PATH, "inference/convert_composer_to_hf.py",
        "--composer_path", composer_checkpoint_path,
        "--hf_output_path", hf_output_path,
        "--output_precision", "bf16"
    ]
    result = subprocess.run(convert_cmd, capture_output=True, text=True)
    print(result.stdout)
    if result.stderr:
        print("Conversion errors:", result.stderr)
    MODEL_CHECKPOINT_VOLUME.commit()
    print("Conversion complete!")


@app.function(gpu=TRAINING_GPU, image=image, timeout=3600, secrets=[Secret.from_name("LRG")],
              volumes={MODEL_CHECKPOINT_VOLUME_MOUNT_PATH: MODEL_CHECKPOINT_VOLUME},
              concurrency_limit=1)
def evaluate_model(checkpoint_path: str):
    import subprocess, os
    from pathlib import Path

    os.chdir("/llm-foundry/scripts")
    print(f"Working directory: {os.getcwd()}")
    
    model_path = Path(MODEL_CHECKPOINT_VOLUME_MOUNT_PATH)/checkpoint_path
    save_path = model_path.parent/"evals"  # Create evals subfolder path
    
    print("\nEvaluating model...")
    eval_cmd = [
        "composer",
        "eval/eval.py",
        "eval/yamls/hf_eval.yaml",
        "icl_tasks=eval/yamls/copa.yaml",
        f"variables.model_name_or_path={model_path}",
        f"results_path={save_path}"  # Add results_path parameter
    ]
    result = subprocess.run(eval_cmd, capture_output=True, text=True)
    print(result.stdout)
    if result.stderr:
        print("Evaluation errors:", result.stderr)
    
    MODEL_CHECKPOINT_VOLUME.commit()  # Commit the new eval results
    print("Evaluation complete!")


@app.function(gpu=TRAINING_GPU, image=image, timeout=3600, secrets=[Secret.from_name("LRG")],
              volumes={MODEL_CHECKPOINT_VOLUME_MOUNT_PATH: MODEL_CHECKPOINT_VOLUME},
              concurrency_limit=1)
def generate_responses(checkpoint_path: str, prompts: list[str]|str|None=None):
    import subprocess, os
    from pathlib import Path

    os.chdir("/llm-foundry/scripts")
    print(f"Working directory: {os.getcwd()}")
    
    model_path = Path(MODEL_CHECKPOINT_VOLUME_MOUNT_PATH)/checkpoint_path

    if prompts is None:
        prompts = [
            "The answer to life, the universe, and happiness is",
            "Here's a quick recipe for baking chocolate chip cookies: Start by"
        ]
    elif isinstance(prompts, str):
        prompts = [prompts]
    

    print("\nGenerating test responses...")
    generate_cmd = [
        PYTHON_PATH, "inference/hf_generate.py",
        "--name_or_path", model_path,
        "--max_new_tokens", "256",
        "--prompts",
        *prompts
    ]
    result = subprocess.run(generate_cmd, capture_output=True, text=True)
    print(result.stdout)
    if result.stderr:
        print("Generation errors:", result.stderr)
    print("Generation complete!")


@app.local_entrypoint()
def main():
    from pathlib import Path
    import time
    run_ts = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")

    get_stats.remote()
    time.sleep(1)
    # convert_c4_small_dataset.remote() # Only run once

    model_path = train_with_aim.remote(run_ts)
    print(f"Model path: {model_path}")
    model_path = Path(model_path).name
    time.sleep(1)
    
    # view_model_checkpoints.remote()
    # time.sleep(1)
    # convert_model_to_hf.remote(model_path)
    # time.sleep(1)
    # evaluate_model.remote(model_path+"/hf")
    # time.sleep(1)
    # generate_responses.remote(model_path+"/hf")
```

**Comment by galopyz** [comment]
@vishalbakshi I ran 2 more times for 100 batch with FA2, 1bs and got 8:33 ish for both.

I am not sure what happened with previous run. I checked the run params on AIM and they were the same. Maybe I was crazy, or maybe the system had a glitch. I should run couple times for each run and take the average or something to make sure it is right.

**Comment by vishalbakshi** [comment]
Sounds good! I'm running some as well. I did runs without eval and the times look reasonable (30 sec for 100ba, 107 sec for 500ba):

![Image](https://github.com/user-attachments/assets/474a3f5e-c8ce-43a6-8530-72f35d656893)

Doing it with eval=100ba and eval=500ba next.

**Comment by vishalbakshi** [comment]
Here are my results---the evaluation step is a massive time sink, which is why 100ba and 500ba take such similar times:

![Image](https://github.com/user-attachments/assets/24dd01d6-1659-4977-811d-dfc7a0b6b798)

**Comment by galopyz** [comment]
FA2 speeds up! Let's use it 👍
